### PR TITLE
request: fix intermittent race when verbose and reading the body in different routines

### PR DESCRIPTION
**Pull request by shazbert**
# PR Description

Noted as issue by @cranktakular THANKS MATE. 

Fixes # (issue)

* race when doRequest is called and the io.Reader as request.Body is sent to the log package which is deferred to another routine. 

```
WARNING: DATA RACE
Write at 0x00c000624438 by goroutine 29:
  bytes.(*Buffer).Read()
      /Go/src/bytes/buffer.go:329 +0x1c4
  io.(*nopCloserWriterTo).Read()
      <autogenerated>:1 +0x6c
  net/http.(*readTrackingBody).Read()
      /Go/src/net/http/transport.go:657 +0x76
  net/http.(*http2clientStream).writeRequestBody()
      /Go/src/net/http/h2_bundle.go:8760 +0x5d9
  net/http.(*http2clientStream).writeRequest()
      /Go/src/net/http/h2_bundle.go:8467 +0xf32
  net/http.(*http2clientStream).doRequest()
      /Go/src/net/http/h2_bundle.go:8353 +0x33
  net/http.(*http2ClientConn).RoundTrip.gowrap1()
     /Go/src/net/http/h2_bundle.go:8259 +0x44

Previous read at 0x00c000624438 by goroutine 7:
  bytes.(*Buffer).String()
      /Go/src/bytes/buffer.go:71 +0x3a
  fmt.(*pp).handleMethods()
      /Go/src/fmt/print.go:673 +0x6ea
  fmt.(*pp).printArg()
     /Go/src/fmt/print.go:756 +0xb4b
  fmt.(*pp).doPrintf()
     /Go/src/fmt/print.go:1075 +0x592
  fmt.Sprintf()
      /Go/src/fmt/print.go:239 +0x5c
  github.com/thrasher-corp/gocryptotrader/log.Debugf.(*fields).stagef.func1()
     /gocryptotrader/log/loggers.go:274 +0x64
  github.com/thrasher-corp/gocryptotrader/log.loggerWorker()
      /gocryptotrader/log/logger_multiwriter.go:38 +0x155
```

## Type of change

Please delete options that are not relevant and add an `x` in `[]` as item is complete.

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## How has this been tested

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration and
also consider improving test coverage whilst working on a certain feature or package.

- [ ] go test ./... -race
- [ ] golangci-lint run
- [ ] Test X

## Checklist

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation and regenerated documentation via the documentation tool
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally and on Github Actions with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: exchanges/request/request.go [hunk 1]
```diff
@@ -172,7 +173,15 @@ func (r *Requester) doRequest(ctx context.Context, endpoint EndpointLimit, newRe
...
-				log.Debugf(log.RequestSys, "%s request body: %v", r.name, p.Body)
+				var data []byte
+				// Below reads the body and then resets it for the actual request.
+				data, err = io.ReadAll(p.Body)
+				if err != nil {
+					log.Errorf(log.RequestSys, "%s failed to read request body: %v", r.name, err)
+				} else {
+					p.Body = bytes.NewBuffer(data)
+				}
+				log.Debugf(log.RequestSys, "%s request body: %v", r.name, string(data))
```
Comment (gloriousCode) [inline exchanges/request/request.go:184]: Would you consider this? Si that you don't need to reset the body again?

```diff
diff --git a/exchanges/request/request.go b/exchanges/request/request.go
index a2cb626b5..bf8973644 100644
--- a/exchanges/request/request.go
+++ b/exchanges/request/request.go
@@ -1,7 +1,6 @@
 package request
 
 import (
-       "bytes"
        "context"
        "encoding/json"
        "errors"
@@ -166,22 +165,22 @@ func (r *Requester) doRequest(ctx context.Context, endpoint EndpointLimit, newRe
 
                verbose := isVerbose(ctx, p.Verbose)
 
                        log.Debugf(log.RequestSys, "%s attempt %d request path: %s", r.name, attempt, p.Path)
                        for k, d := range req.Header {
                                log.Debugf(log.RequestSys, "%s request header [%s]: %s", r.name, k, d)
                        }
                        log.Debugf(log.RequestSys, "%s request type: %s", r.name, p.Method)
-                       if p.Body != nil {
-                               var data []byte
-                               // Below reads the body and then resets it for the actual request.
-                               data, err = io.ReadAll(p.Body)
-                               if err != nil {
-                                       log.Errorf(log.RequestSys, "%s failed to read request body: %v", r.name, err)
-                               } else {
-                                       p.Body = bytes.NewBuffer(data)
+                       if req.GetBody != nil {
+                               bodyCopy, bodyErr := req.GetBody()
+                               if bodyErr != nil {
+                                       return bodyErr
+                               }
+                               payload, bodyErr := io.ReadAll(bodyCopy)
+                               if bodyErr != nil {
+                                       return bodyErr
                                }
-                               log.Debugf(log.RequestSys, "%s request body: %v", r.name, string(data))
+                               log.Debugf(log.RequestSys, "%s request body: %s", r.name, payload)
                        }
                }
```

**Review by gloriousCode — CHANGES_REQUESTED** [review]
# 🏎️💨💨💨💨💨💨💨💨💨💨💨💨💨💨

File: exchanges/request/request.go (hunk 1 above)
Comment (gloriousCode) [inline exchanges/request/request.go:184]: Would you consider this? Si that you don't need to reset the body again?

```diff
diff --git a/exchanges/request/request.go b/exchanges/request/request.go
index a2cb626b5..bf8973644 100644
--- a/exchanges/request/request.go
+++ b/exchanges/request/request.go
@@ -1,7 +1,6 @@
 package request
 
 import (
-       "bytes"
        "context"
        "encoding/json"
        "errors"
@@ -166,22 +165,22 @@ func (r *Requester) doRequest(ctx context.Context, endpoint EndpointLimit, newRe
 
                verbose := isVerbose(ctx, p.Verbose)
 
                        log.Debugf(log.RequestSys, "%s attempt %d request path: %s", r.name, attempt, p.Path)
                        for k, d := range req.Header {
                                log.Debugf(log.RequestSys, "%s request header [%s]: %s", r.name, k, d)
                        }
                        log.Debugf(log.RequestSys, "%s request type: %s", r.name, p.Method)
-                       if p.Body != nil {
-                               var data []byte
-                               // Below reads the body and then resets it for the actual request.
-                               data, err = io.ReadAll(p.Body)
-                               if err != nil {
-                                       log.Errorf(log.RequestSys, "%s failed to read request body: %v", r.name, err)
-                               } else {
-                                       p.Body = bytes.NewBuffer(data)
+                       if req.GetBody != nil {
+                               bodyCopy, bodyErr := req.GetBody()
+                               if bodyErr != nil {
+                                       return bodyErr
+                               }
+                               payload, bodyErr := io.ReadAll(bodyCopy)
+                               if bodyErr != nil {
+                                       return bodyErr
                                }
-                               log.Debugf(log.RequestSys, "%s request body: %v", r.name, string(data))
+                               log.Debugf(log.RequestSys, "%s request body: %s", r.name, payload)
                        }
                }
```

Output:
```
[DEBUG]  | REQUESTER | 21/03/2024 08:35:09 | BTC Markets attempt 1 request path: https://api.btcmarkets.net/v3/orders
[DEBUG]  | REQUESTER | 21/03/2024 08:35:09 | BTC Markets request header [Bm-Auth-Signature]: [WIF420WIF420WIF420WIF420WIF420WIF420WIF420WIF420WIF420WIF420]
[DEBUG]  | REQUESTER | 21/03/2024 08:35:09 | BTC Markets request header [Accept]: [application/json]
[DEBUG]  | REQUESTER | 21/03/2024 08:35:09 | BTC Markets request header [Accept-Charset]: [UTF-8]
[DEBUG]  | REQUESTER | 21/03/2024 08:35:09 | BTC Markets request header [Content-Type]: [application/json]
[DEBUG]  | REQUESTER | 21/03/2024 08:35:09 | BTC Markets request header [Bm-Auth-Apikey]: [WIF420WIF420WIF420WIF420WIF420WIF420WIF420WIF420WIF420WIF420]
[DEBUG]  | REQUESTER | 21/03/2024 08:35:09 | BTC Markets request header [Bm-Auth-Timestamp]: [1710970509087]
[DEBUG]  | REQUESTER | 21/03/2024 08:35:09 | BTC Markets request type: POST
[DEBUG]  | REQUESTER | 21/03/2024 08:35:09 | BTC Markets request body: {"amount":"1","marketId":"BTC-AUD","postOnly":true,"price":"100","side":"Bid","type":"Limit"}
```

**Review by gloriousCode — APPROVED** [review]
Good thinking with closing the reader... 😄 
Nice work programmers @cranktakular & @shazbert  tACK

**Review by thrasher- — APPROVED** [review]
Nice catch @cranktakular ! Thanks @shazbert

**Pull request merged**